content/posts/2021/reana-client-0.7.3.md

```diff
@@ -0,0 +1,346 @@
+         commands:
+-          - python "${script_path}" "${sleptime}"
++          - python "${script_path}" "${sleeptime}"
+```
+
+```console
+$ reana-client validate -f reana.yaml
+==> Verifying REANA specification file... my-analysis/reana.yaml
+  -> SUCCESS: Valid REANA specification file.
+==> Verifying workflow parameters and commands...
+```
+
+Now everything is settled, the warnings disappear as the new
+parameter is defined and used properly.
+
+This validation is also compatible with Yadage and CWL workflows,
+where more advanced verifications are carried out as parameters
+can be defined at multiple levels.
+
+### Workflow dangerous operations verification
+
+Over time, we have identified certain commands or operations that
+can potentially cause problems when the workflows are run in the
+REANA Cluster. It is our mission to detect these kinds of operations
+and warn users that their execution might cause errors.
+
+Some of these operations are running commands as the system superuser
+(`sudo`) or changing the working directory where the workflow is run
+instead of using the workflow workspace.
+
+Let us modify our previous example to execute the Python script as
+the superuser, which we strongly discourage as the user running the
+workflow might not have the rights to do this:
+
+```diff
+       - name: run-script
+         environment: 'python:3.9-slim'
+         commands:
+-          - python "${script_path}" "${sleeptime}"
++          - sudo python "${script_path}" "${sleeptime}"
+```
+
+Let us run the validator:
+
+```console
+$ reana-client validate -f reana.yaml
+==> Verifying REANA specification file... my-analysis/reana.yaml
+  -> SUCCESS: Valid REANA specification file.
+==> Verifying workflow parameters and commands...
+  -> WARNING: Operation "sudo" found in step "run-script" might be dangerous.
+```
+
+The output shows a warning message indicating the dangerous operation
+and in which step was found.
+
+### Workflow environment images validation
+
+Workflows run in containerised environments that can evolve
+throughout versions via tags. We encourage users to tag their images
+when running analyses as this will ensure reproducibility in the
+future, one of the pillars of
+[FAIR principles](https://www.nature.com/articles/sdata201618).
+
+Besides, it is important to verify the existence of these
+environment images, both locally and remotely (Docker Hub and
+GitLab registry), to ensure that the REANA cluster can pull them with
+no issues.
+
+This is why we are introducing a new option for the `validate`
+command, called `--environments`. When passing this option to the
+validator, we can perform the aforementioned verifications.
+
+Let us use the previous workflow example to depict this
+functionality. In that workflow, the environment used was
+`python:3.9-slim`, which is properly tagged and exists in
+Docker Hub. We are going to modify it and use a non-existing image
+instead, for example `foo:bar`.
```

Comment:
Instead of `foo:bar`, what about again using a typo technique` phyton:3.9-slim`.  Advantage: this will be more natural transition to `python:latest` used below.

Alternatively, we could even start with latest tag, so that the example thinking line will go like:

- python:3.9-slim -- previous example
- python:latest -- tag troubles
- python -- tag troubles
- phyton -- typo, existence troubles